implementations/rust/ockam/ockam_node/src/storage/database/sqlx_types.rs

```diff
@@ -0,0 +1,200 @@
+    /// This type represents floats
+    #[allow(unused)]
+    Real(f64),
+}
+
+/// The SqlType implements the Type<Sqlite> trait from sqlx to allow its values to be serialized
+/// to an Sqlite database
+impl Type<Sqlite> for SqlxType {
+    fn type_info() -> <Sqlite as Database>::TypeInfo {
+        <Vec<u8> as Type<Sqlite>>::type_info()
+    }
+}
+
+/// The SqlType implements the Encode<Sqlite> trait from sqlx to allow its values to be serialized
+/// to an Sqlite database. There is a 1 to 1 mapping with the database native types
+impl Encode<'_, Sqlite> for SqlxType {
+    fn encode_by_ref(&self, buf: &mut <Sqlite as HasArguments>::ArgumentBuffer) -> IsNull {
+        match self {
+            SqlxType::Text(v) => <String as Encode<'_, Sqlite>>::encode_by_ref(v, buf),
+            SqlxType::Blob(v) => <Vec<u8> as Encode<'_, Sqlite>>::encode_by_ref(v, buf),
+            SqlxType::Integer(v) => <i64 as Encode<'_, Sqlite>>::encode_by_ref(v, buf),
+            SqlxType::Real(v) => <f64 as Encode<'_, Sqlite>>::encode_by_ref(v, buf),
+        }
+    }
+
+    fn produces(&self) -> Option<<Sqlite as Database>::TypeInfo> {
+        Some(match self {
+            SqlxType::Text(_) => <String as Type<Sqlite>>::type_info(),
+            SqlxType::Blob(_) => <Vec<u8> as Type<Sqlite>>::type_info(),
+            SqlxType::Integer(_) => <i64 as Type<Sqlite>>::type_info(),
+            SqlxType::Real(_) => <f64 as Type<Sqlite>>::type_info(),
+        })
+    }
+}
+
+/// This trait can be implemented by any type that can be converted to a database type
+/// Typically an `Identifier` (to a `Text`), a `TimestampInSeconds` (to an `Integer`) etc...
+///
+/// This allows a value to be used as a bind parameters in a sqlx query for example:
+///
+/// use std::str::FromStr;
+/// use sqlx::query_as;
+/// use ockam_node::database::{SqlxType, ToSqlxType};
+///
+/// // newtype for a UNIX-like timestamp
+/// struct TimestampInSeconds(u64);
+///
+/// // this implementation maps the TimestampInSecond type to one of the types that Sqlx
+/// // can serialize for sqlite
+/// impl ToSqlxType for TimestampInSeconds {
+///     fn to_sql(&self) -> SqlxType {
+///         self.0.to_sql()
+///     }
+/// }
+///
+/// let timestamp = TimestampInSeconds(10000000);
+/// let query = query_as("SELECT * FROM identity WHERE created_at >= $1").bind(timestamp.as_sql());
+///
+///
+pub trait ToSqlxType {
+    /// Return the appropriate sql type
+    fn to_sql(&self) -> SqlxType;
+}
+
+impl ToSqlxType for String {
+    fn to_sql(&self) -> SqlxType {
+        SqlxType::Text(self.clone())
+    }
+}
+
+impl ToSqlxType for &str {
+    fn to_sql(&self) -> SqlxType {
+        self.to_string().to_sql()
+    }
+}
+
+impl ToSqlxType for bool {
+    fn to_sql(&self) -> SqlxType {
+        if *self {
+            1.to_sql()
+        } else {
+            0.to_sql()
+        }
+    }
+}
+
+impl ToSqlxType for u64 {
+    fn to_sql(&self) -> SqlxType {
+        SqlxType::Integer(*self as i64)
```

Comment:
That's potentially unsafe

Reply:
What do you propose to do instead?

Reply:
Let's introduce an unsigned integer type to `ToSqlxType` and test that it indeed is correct for values more than 2^32 after a store/retrieve cycle.
Not sure what to do with a usize, do we consider platforms with more than 64 arch?

Reply:
I don't see what you mean. Can you please add a commit showing your idea?

Reply:
Maybe we can use `i64::try_from(self)` and raise a proper panic message if we ever overflow?

Reply:
I'll need to play with it for a bit today. If that's the only thing left in this PR - let's merge it and create an issue for that

Reply:
Ok, so what actually happens when you have a u64 and its values doesn't fit into i64, is that it gets overflowed (e.g., u64::MAX -> -1i64). Then when you retrieve it from the db, you convert it again and it is overflowed again, and now has the correct value again. I think it's far from ideal, but it works 😅 I don't have a better alternative atm